include/fiction/algorithms/simulation/sidb/determine_displacement_robustness.hpp

```diff
@@ -0,0 +1,542 @@
+
+        std::shuffle(all_possible_sidb_displacement.begin(), all_possible_sidb_displacement.end(), generator);
+
+        std::vector<Lyt> layouts{};
+        layouts.reserve(all_possible_sidb_displacement.size());
+
+        std::size_t number_of_layouts_with_displaced_sidbs = 0;
+
+        double percentage_of_layouts = 1.0;
+        if (params.analysis_mode ==
+            displacement_robustness_domain_params<Lyt>::displacement_analysis_mode::RANDOM_SAMPLING)
+        {
+            percentage_of_layouts = params.percentage_of_analyzed_displaced_layouts;
+        }
+
+        assert(percentage_of_layouts >= 0.0 && percentage_of_layouts <= 1.0 &&
+               "percentage_of_analyzed_displaced_layouts must be between 0.0 and 1.0");
+
+        const auto max_number_of_layouts_with_displaced_sidbs =
+            std::max(static_cast<std::size_t>(2),
+                     static_cast<std::size_t>(percentage_of_layouts *
+                                              static_cast<double>(all_possible_sidb_displacement.size())));
+        // the "2" is used so that at least one further displaced layout is analyzed in addition to the original SiDB
+        // layout.
+
+        for (const auto& cell_displacements : all_possible_sidb_displacement)
+        {
+            if (number_of_layouts_with_displaced_sidbs >= max_number_of_layouts_with_displaced_sidbs)
+            {
+                break;
+            }
+            Lyt         lyt{};
+            std::size_t identical_cells_to_original_layout = 0;
+            for (std::size_t i = 0; i < cell_displacements.size(); ++i)
+            {
+                lyt.assign_cell_type(cell_displacements[i], layout.get_cell_type(sidbs_of_the_original_layout[i]));
+                if (!layout.is_empty_cell(cell_displacements[i]))
+                {
+                    identical_cells_to_original_layout++;
+                }
+            }
+            if (lyt.num_cells() == layout.num_cells() && identical_cells_to_original_layout != layout.num_cells())
+            {
+                layouts.push_back(lyt);
+            }
+            number_of_layouts_with_displaced_sidbs++;
+        }
+        return layouts;
+    }
+    /**
+     * The SiDB cell-level layout/SiDB gate for which the displacement robustness calculation is performed.
+     */
+    const Lyt& layout;
+    /**
+     * The parameters for the displacement robustness computation.
+     */
+    displacement_robustness_domain_params<Lyt> params;
+    /**
+     * The statistics of the displacement robustness computation.
+     */
+    displacement_robustness_domain_stats& stats;
+    /**
+     * This stores all possible displacements for all SiDBs in the SiDB layout. This means e.g. the first vector
+     * describes all possible positions of the first SiDB due to the allowed/possible displacements.
+     */
+    std::vector<std::vector<cell<Lyt>>> all_possible_displacements_for_each_sidb{};
```

Comment:
Maybe a map from `cell<Lyt>` to `std::vector<cell<Lyt>>` is a more appropriate data structure?